storage/listv2.go

```diff
@@ -0,0 +1,94 @@
+// Copyright (C) 2018 Storj Labs, Inc.
+// See LICENSE for copying information.
+
+package storage
+
+import "errors"
+
+// More indicates if the result was truncated. If false
+// then the result []ListItem includes all requested keys.
+// If true then the caller must call List again to get more
+// results by setting `StartAfter` or `EndBefore` appropriately.
+type More bool
+
+// ListOptions are items that are optional for the LIST method
+type ListOptions struct {
+	Prefix       Key
+	StartAfter   Key
+	EndBefore    Key
+	Recursive    bool
+	IncludeValue bool
+	Limit        Limit
+}
+
+// ListV2 lists all keys corresponding to ListOptions
+func ListV2(store KeyValueStore, opts ListOptions) (result Items, more More, err error) {
+	if opts.StartAfter != nil && opts.EndBefore != nil {
+		return nil, false, errors.New("start-after and end-before cannot be combined")
+	}
+
+	reverse := opts.EndBefore != nil
+
+	more = More(true)
+	limit := opts.Limit
+	if limit == 0 {
+		limit = Limit(1 << 31)
```

Comment:
my preference here would be to return an invalid request error (long term, maybe that won't pass tests right now)

Reply:
I'll change that behavior in a next PR. Would disrupt other packages.

Reply:
Why not having an internal limit if the client does not care about a limit ifself? Of course it should be a more sane value, e.g. 1000, not 1 << 31.